### ISSUE TITLE
STI is not extendable

### ISSUE THREAD

**Issue by mfojtik** [openshift/source-to-image#122]
Rightnow  the STI source code is very 'sti-centric'. If I want to add my custom build type (heroku build-pack, generic Docker build, etc...) I can't easily plug this into the existing STI codebase. We should have better interfaces and have factories for this:

```
type Builder interface {
  Build()
}

type Downloader interface {
  Download()
}
```

What we have right now is something like:

```
type buildHandlerInterface interface {
	cleanup()
	setup(required []api.Script, optional []api.Script) error
	determineIncremental() error
	Request() *api.Request
	Result() *api.Result
	saveArtifacts() error
	fetchSource() error
	execute(command api.Script) error
	wasExpectedError(text string) bool
	build() error
}
```

and so on... the STI will be still 'default' build type, but we can also support different build models in the future.

This issue requires massive refactoring, so I don't think we can make it real before GA, but I want to keep this issue opened here that we don't forget about it.

**Comment by mfojtik** [comment]
@soltysh @bparees fyi ^^

**Comment by bparees** [comment]
+1

**Comment by kargakis** [comment]
I am still not very familiar with the STI codebase but this interface design feels quite rough with all the methods stacked into one interface.

tl;dr +1, a more composable interface design will probably be better

**Comment by mfojtik** [comment]
This is some very rough attempt to sneak build 'strategies' into STI:

https://github.com/mfojtik/source-to-image/commit/faec143c260d4a0f38c94db2a54226a1cea48bf1

and it actually works:

https://gist.github.com/mfojtik/a3b1504f6e222ee38db2

**Comment by mfojtik** [comment]
Fixed.
